### ISSUE TITLE
Invalid class matches in analyzer.

### ISSUE THREAD

**Issue by SandPod** [serverpod/serverpod#3019]
**Describe the bug**
This is a theoretical issue that should be addressed.

In a recent PR, https://github.com/serverpod/serverpod/pull/3018, an issue was fixed that could result in the wrong class definition being fetched from the parsed class list.

When searching the repository, there are more places that might result in the same issue.

It is unclear what implications these issue might have on the runtime of Serverpod.

A search after `.className ==` reveals these locations with this potential issue:

- [ ] https://github.com/serverpod/serverpod/blob/0009d80b52417669e80dd0a728fc00dae409bcea/tools/serverpod_cli/lib/src/analyzer/models/entity_dependency_resolver.dart#L47
- [ ] https://github.com/serverpod/serverpod/blob/0009d80b52417669e80dd0a728fc00dae409bcea/tools/serverpod_cli/lib/src/analyzer/models/entity_dependency_resolver.dart#L96
- [ ] https://github.com/serverpod/serverpod/blob/0009d80b52417669e80dd0a728fc00dae409bcea/tools/serverpod_cli/lib/src/analyzer/models/entity_dependency_resolver.dart#L114
- [ ] https://github.com/serverpod/serverpod/blob/0009d80b52417669e80dd0a728fc00dae409bcea/tools/serverpod_cli/lib/src/analyzer/models/entity_dependency_resolver.dart#L350
- [ ] https://github.com/serverpod/serverpod/blob/0009d80b52417669e80dd0a728fc00dae409bcea/tools/serverpod_cli/lib/src/analyzer/models/validation/restrictions.dart#L172
- [ ] https://github.com/serverpod/serverpod/blob/0009d80b52417669e80dd0a728fc00dae409bcea/tools/serverpod_cli/lib/src/analyzer/models/validation/restrictions.dart#L1415
- [ ] https://github.com/serverpod/serverpod/blob/0009d80b52417669e80dd0a728fc00dae409bcea/tools/serverpod_cli/lib/src/generator/types.dart#L535 

These places should be patched and a test should be added that validate the correct class is fetched.

**Additional context**
Some users have had some strange behavior when generating the Serverpod project. I don't have a discussion thread to link to right now, but I can swear I have seen similar issues.

**Comment by Isakdl** [comment]
I don't think this should be an issue, we already have a rule that will give you an error if there are multiple classes with the same name, this will prevent the code generator from creating these classes, hence you will never get a runtime error.

And how can you know which class is the correct one? The only information you have is the class name (type) and if that is match with several entries you cannot differentiate between them, that is an error in it self and as I said is already covered!

**Comment by SandPod** [comment]
> I don't think this should be an issue, we already have a rule that will give you an error if there are multiple classes with the same name, this will prevent the code generator from creating these classes, hence you will never get a runtime error.
> 
> And how can you know which class is the correct one? The only information you have is the class name (type) and if that is match with several entries you cannot differentiate between them, that is an error in it self and as I said is already covered!

This is fine as long as you are working with a single project, but we do allow the same class to be defined in multiple modules. The way they are differentiated then is by the `moduleAlias`.

**Comment by SandPod** [comment]
A bug from this was probably reported here: https://github.com/serverpod/serverpod/issues/2857
